examples/rcn.py

```diff
@@ -0,0 +1,438 @@
+from sklearn.datasets import fetch_openml
+
+from pgmax.fg import graph, groups
+
+fetch_openml_cached = memory.cache(fetch_openml)
+os.environ["XLA_PYTHON_CLIENT_ALLOCATOR"] = "platform"
+os.environ["XLA_PYTHON_CLIENT_PREALLOCATE"] = "false"
+
+# %% [markdown]
+# # 1. Load the data
+# %% [markdown]
+# In this notebook, we train an RCN model on a small subset of MNIST (training on 20 images and testing on 20 images). We also have support to train the model on 100 images. Use this option only if you have a gpu with atleast 8Gbs memory. Also, recommend that jax is installed with cuda enabled for training with more than 20 images.
+
+# %%
+train_size = 20
+test_size = 20
+
+
+# %%
+def fetch_mnist_dataset(test_size: int, seed: int = 5) -> Tuple[np.ndarray, np.ndarray]:
+    """Returns test images sampled randomly from the set of MNIST images.
+
+    Args:
+        test_size: Desired number of test images.
+        seed: An integer used for repeatable tests.
+
+    Returns:
+        test_set: A list of length test_size containing images from the MNIST test dataset.
+        test_labels: Corresponding labels for the test images.
+    """
+
+    mnist_train_size = 60000
+    num_per_class = test_size // 10
+
+    print("Fetching the MNIST dataset")
+    dataset = fetch_openml_cached("mnist_784", as_frame=False, cache=True)
+    print("Successfully downloaded the MNIST dataset")
+
+    mnist_images = dataset["data"]
+    mnist_labels = dataset["target"].astype("int")
+    full_mnist_test_images = mnist_images[mnist_train_size:]
+    full_mnist_test_labels = mnist_labels[mnist_train_size:]
+
+    np.random.seed(seed)
+
+    test_set = []
+    test_labels = []
+    for i in range(10):
+        idxs = np.random.choice(
+            np.argwhere(full_mnist_test_labels == i)[:, 0], num_per_class
+        )
+        for idx in idxs:
+            img = full_mnist_test_images[idx].reshape(28, 28)
+            img_arr = jax.image.resize(image=img, shape=(112, 112), method="bicubic")
+            img = jnp.pad(
+                img_arr,
+                pad_width=tuple([(p, p) for p in (44, 44)]),
+                mode="constant",
+                constant_values=0,
+            )
+
+            test_set.append(img)
+            test_labels.append(i)
+
+    return np.array(test_set), np.array(test_labels)
+
+
+# %%
+test_set, test_labels = fetch_mnist_dataset(test_size)
+
+# %% [markdown]
+# # 2. Load the model
+
+# %%
+data = np.load("example_data/rcn.npz", allow_pickle=True, encoding="latin1")
+idxs = range(0, 100, 100 // train_size)
+
+train_set, train_labels, frcs, edges, suppression_masks, filters = (
+    data["train_set"][idxs, :, :],
+    data["train_labels"][idxs],
+    data["frcs"][idxs],
+    data["edges"][idxs],
+    data["suppression_masks"],
+    data["filters"],
+)
+
+hps, vps = 12, 12  # Horizontal and vertical pool sizes respectively for RCN models.
+num_orients = filters.shape[0]
+brightness_diff_threshold = 40.0
+max_perturb_radius = 22
```

Comment:
I would re-group them and explain every variable defined here, with either an inline comment or a comment above it, like you did for pool sizes.
num_orients and bdt are related to preproc to run edge detection.
different from the core PGM model.

Reply:
It also won't hurt to expand the pool size to indicate what it exactly is, is it half the window size, is it radius or full size?

Reply:
Done.